Show course home messages for important course dates

andy-armstrong opened the pull request:
## [LEARNER-2073](https://openedx.atlassian.net/browse/LEARNER-2073)

### Description

This change generalizes the logic for showing messages for important course dates. Now any course date class can register an alert, and I've updated the start date message to use this framework. I've then implemented messages for end dates and verification dates.

![image](https://user-images.githubusercontent.com/5985072/30840098-6db72af0-a243-11e7-9bec-2f8f722ebf60.png)

### Acceptance Criteria

- [x] Show a course home message for important course dates
  - [x] Covers Upgrade Deadline, and Course End
  - [x] Some messages will have a link (upgrade, id verify)
  - [ ] Add a unique class per message type in support of enabling/disabling in Optimizely
  - [x] Add an event for each link (bi?)
- [x] If multiple milestones are relevant:
  - [x] show a separate message for each one
  - [x] sort by date (earliest milestones shown first)
- [x] Milestones will be warned two weeks ahead of their date
  - [x] For very short courses (under 2 weeks), can we start with default behavior that could include   
  - [x] multiple messages at the same time as the pre-course start message?
- [x] Note: none of these messages will be dismissable

### Sandbox

- https://andy-armstrong.sandbox.edx.org/courses/course-v1:AndyA+VERIFIED+1/course/
- https://andy-armstrong.sandbox.edx.org/courses/course-v1:AndyA+FUTURE+1/course/

### Testing
- [x] Unit, integration, acceptance tests as appropriate
- [x] Analytics
- [x] Performance
- [x] Database migrations are backwards-compatible
- [x] safecommit violation code review process

Front-end changes:
- [x] Accessibility (Check for a11y violations, ensure keyboard accessible, screen reader testing as appropriate)
- [x] i18n
- [x] RTL

### Assign Github reviewers (as needed)
- [x] engineering 
- [x] product
- [ ] UX
- [x] documentation
- [ ] a11y, if includes custom UI (Alasdair Swan is the learner team a11y lead)

FYI: @edx/learner-mercury
 
### Post-review
- [x] Rebase and squash commits

robrap commented:
@marcore @andy-armstrong: Do we need to avoid having the upgrade message appear at the same time as the current hero?

robrap commented:
@andy-armstrong: Let me know if you want to discuss eventing.  We should add the ecommerce events and I can show you want we want, but not sure where that should live.

marcotuts commented:
Other than adding a waffle flag to keep the upgrade message disabled until after we complete the hero sidebar testing, this looks good to me! (Eventing I know is still in progress, but mostly referring to product acceptance type review.)

katymyw commented:
@marcotuts -- It sounds like the upgrade date messaging is being waffle flagged off for now. Just wanted to flag then if/when we turn that on, we should touch base about the messaging. For the upgrade 'deadlines' in self-paced we are changing some of the language so that the messaging is softer, since it's not a true deadline the same way it is in instructor-paced.

andy-armstrong commented:
@HarryRein @marcotuts @grantgoodmanedX this PR is ready for review now, and I've assigned the three of you as reviewers. I realize that the hackathon is happening so no worries if it has to wait until Friday. I've completed the unit tests now which verify the behavior for all sorts of durations (5 minutes to go, 1 hour to go, 1 day to go, 2 weeks to go, etc).

There are a couple of test regressions that I need to look into, but they shouldn't affect the review.

HarryRein commented on lms/djangoapps/courseware/date_summary.py:
```diff
@@ -41,6 +46,12 @@ def description(self):
+        Registers any relevant course alerts given the current request.
+        """
+        pass
+
```
^this is awesome!

HarryRein commented on lms/djangoapps/courseware/date_summary.py:
```diff
@@ -195,6 +240,39 @@ def description(self):
+        """
+        Registers an alert if the end date is approaching.
+        """
+        now = datetime.datetime.now(UTC())
```
For some reason the course end date in the message and in the right sidebar are different, I believe because in this case the course end date was set to 12am and there is a slight issue with timezones.

HarryRein commented:
I'm not seeing the 'Course Start Date or 'Upgrade to Verified' messages on the sandbox, but am seeing the 'Course end Date' message - any ideas?

HarryRein commented on lms/djangoapps/courseware/date_summary.py:
```diff
@@ -299,6 +371,47 @@ def date(self):
+                    ).format(
+                        upgrade_url=self.link,
+                        upgrade_label=Text(_('Upgrade ({upgrade_price})')).format(
+                            upgrade_price='${upgrade_price}'.format(upgrade_price=upgrade_price),
```
Use format_course_price from course_mode.models to make this i18n compliant

HarryRein commented on lms/static/sass/features/_course-experience.scss:
```diff
@@ -16,6 +16,7 @@
   .message-content {
     @include margin(0, 0, $baseline, $baseline);
+
     position: relative;
```
No change needed but curious why you added this line - do you think we should put lines between include/extends statements moving forward?

andy-armstrong replied:
Styleline made me :-)

robrap commented on openedx/features/course_experience/views/course_home_messages.py:
```diff
@@ -122,81 +128,100 @@ def _register_course_home_messages(request, course_id, user_access, course_start
+        request,
+        HTML('{goal_choices_html}{closing_tag}').format(
+            goal_choices_html=goal_choices_html,
+            closing_tag=HTML('</div>')
```
I "think" the closing tag is already included in `goal_choices_html` above where it belongs.  So maybe this could be?:
```
CourseHomeMessages.register_info_message(
    request,
    goal_choices_html,
    title=Text(_('Welcome to {course_display_name}')).format(
        course_display_name=course.display_name
    )
```

andy-armstrong replied:
I believe that this is right, @robrap. I've made the change and we'll see how it looks.

andy-armstrong commented:
@HarryRein Thanks for noticing the issues with the messages not showing up. The start date was a bug, as I changed it to obey the same 14 day limit as the other messages, when in fact it should always show up. The verification date had passed which was why that message didn't show up, so I've pushed it out a week, and so you can now see it. I've addressed all of your other comments except for the time zone issue which I need to dig into further (good catch!).

andy-armstrong commented:
@HarryRein I've updated the sandbox with the latest code, btw.

marcotuts commented:
Is it a configuration issues / gap that makes the upgrade buttons not link to the sandbox equivalent of this: "https://ecommerce.edx.org/basket/"  The sandbox currently links to an older upgrade page we should be skipping to head straight to the checkout flow.

andy-armstrong commented:
@HarryRein I've fixed the issue with the time zones. The problem was that the server doesn't know the browser's suggested time zone, while the dates sidebar was rendering client-side using moment. I've changed the messages to no longer render the formatted date but instead to output the same spans as the dates sidebar. I've updated the sandbox and verified that the problem has gone away.

andy-armstrong commented:
@HarryRein Of course, this has broken a whole bunch of tests! I'll work on getting them fixed on Monday.

andy-armstrong commented:
@marcotuts I'm using the same logic for generating the links for the upgrade buttons as the hero. My suspicion is that it is a sandbox configuration issue as you suggest. I'll look into it on Monday.

andy-armstrong commented:
@dianakhuang I rebased this PR and have broken the eventing tests for JS. Would you be able to look into what happened? Thanks

andy-armstrong commented:
jenkins run bokchoy

andy-armstrong commented:
@HarryRein This is ready for your review when you have a chance.

HarryRein commented on lms/djangoapps/courseware/date_summary.py:
```diff
@@ -400,9 +398,7 @@ def register_alerts(self, request, course):
-                        upgrade_label=Text(_('Upgrade ({upgrade_price})')).format(
-                            upgrade_price='${upgrade_price}'.format(upgrade_price=upgrade_price),
-                        ),
+                        upgrade_label=Text(_('Upgrade ({upgrade_price})')).format(upgrade_price=upgrade_price),
```
Nit: cosmetic_verified_display_price returns a translated string so don't need the _( here

andy-armstrong commented:
jenkins run lettuce
